feat!: The Great Wallet Refactor

Thunkar opened the pull request:
Second part of the `Wallet` refactor that moves us closer to a fully fledged `Wallet SDK` and PXEless apps. Focuses on completely unbinding the `Wallet` interface from a particular `AztecAddress`. The former now refers to the wallet *software*, which exposes a clear API to apps. This API is not yet finalized!

`ContractFunctionInteraction` and friends have been put on a diet, and most of their responsibilities are now the Wallet's

As a consequence of this PR, every single e2e test, the playground, the cli-wallet and boxes have been altered. It's not pretty (yet), but it will be, I promise.

benesjan commented on aztec-up/test/amm_flow.sh:
```diff
@@ -52,7 +52,6 @@ aztec-wallet \
 
 aztec-wallet \
   register-contract $canonical_sponsored_fpc_address SponsoredFPC \
-  -f accounts:main \
```
Why this change?

Thunkar replied:
You never really register a contract for an account, but rather in the wallet's PXE and all accounts can use it. If a wallet wants to implement an auth mechanism so certain contracts cannot access certain account's data, well...that's up to them

benesjan commented on boxes/boxes/react/webpack.config.js:
```diff
@@ -49,6 +49,7 @@ export default (_, argv) => ({
       path: false,
       tty: false,
       url: false,
+      net: false,
```
What's this?

Thunkar replied:
A polyfill required due to the interim import of `test-wallet` in the boxes. I cannot wait to kill them

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -4,71 +4,98 @@ import {
-export class EmbeddedWallet {
-  private pxe!: PXE;
-  connectedAccount: AccountWallet | null = null;
+// This is a minimal implementation of an Aztec wallet
```
It's just weird that it mentions Aztec wallet but then it extends BaseWallet that implements a Wallet. There doesn't seem to be anything called Aztec wallet in the codebase.

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -4,71 +4,98 @@ import {
+    address: AztecAddress
+  ): Promise<Account> {
+    let account: Account | undefined;
+    if (address.equals(AztecAddress.ZERO)) {
```
I would split the function into `createAccount` and `getAccount`.

Having these ifs hidden in the function implementation that make it do something kind of random just result in bad code clarity. Fetching by invalid address should throw an error.

Thunkar replied:
This is not random.

`from` is a mandatory param, but sometimes there's really no `from`...for example when deploying your own account contract, which does an `universalDeployment`. Essentially this gives the wallet the ability to send "anonymous" txs (in the sense there's no account contract involved)

benesjan reviewed (commented):
I just got stuck trying to figure out what the `getFakeAccountDataFor` on EmbeddedWallet does as I don't know what exactly is a fake account and what is it used for.

Maybe I just don't have enough context as it's been a while I worked on `Aztec.js` so I'll pass the review of this to someone else. Can come back to it if you document this.

benesjan reviewed (approved):
As agreed in the team call, what is on `next` is not reasonable and such a huge change needs to get in quickly so I am approving without reviewing.

Happy merging 😊

Thunkar committed:
feat!: The Great Wallet Refactor

Second part of the `Wallet` refactor that moves us closer to a fully fledged `Wallet SDK` and PXEless apps. Focuses on completely unbinding the `Wallet` interface from a particular `AztecAddress`. The former now refers to the wallet *software*, which exposes a clear API to apps. This API is not yet finalized!

`ContractFunctionInteraction` and friends have been put on a diet, and most of their responsibilities are now the Wallet's

As a consequence of this PR, every single e2e test, the playground, the cli-wallet and boxes have been altered. It's not pretty (yet), but it will be, I promise.

Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: David Banks <47112877+dbanks12@users.noreply.github.com>
Co-authored-by: Facundo <fcarreiro@users.noreply.github.com>
Co-authored-by: Gregorio Juliana <gregojquiros@gmail.com>
Co-authored-by: Harsh Bajpai <bajpaiharsh244@gmail.com>
Co-authored-by: Ilyas Ridhuan <ilyas@aztecprotocol.com>
Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: José Pedro Sousa <jose@aztecprotocol.com>
Co-authored-by: Lasse Herskind <16536249+LHerskind@users.noreply.github.com>
Co-authored-by: Leila Wang <leizciw@gmail.com>
Co-authored-by: Lin Oshitani <linoshitani@gmail.com>
Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
Co-authored-by: Maddiaa <47148561+Maddiaa0@users.noreply.github.com>
Co-authored-by: Michael Connor <mike@aztecprotocol.com>
Co-authored-by: Miranda Wood <miranda@aztecprotocol.com>
Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>
Co-authored-by: Paperclip Minimizer <minim@wonderland.xyz>
Co-authored-by: PhilWindle <60546371+PhilWindle@users.noreply.github.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: Santiago Palladino <santiago@aztecprotocol.com>
Co-authored-by: Sarkoxed <75146596+Sarkoxed@users.noreply.github.com>
Co-authored-by: Savio <72797635+Savio-Sou@users.noreply.github.com>
Co-authored-by: Zachary James Williamson <blorktronics@gmail.com>
Co-authored-by: danielntmd <162406516+danielntmd@users.noreply.github.com>
Co-authored-by: guipublic <47281315+guipublic@users.noreply.github.com>
Co-authored-by: josh crites <critesjosh@gmail.com>
Co-authored-by: just-mitch <68168980+just-mitch@users.noreply.github.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: maramihali <mara@aztec-labs.com>
Co-authored-by: mralj <nikola.mratinic@gmail.com>
Co-authored-by: spypsy <spypsy@users.noreply.github.com>
Co-authored-by: thunkar <gregjquiros@gmail.com>
Co-authored-by: thunkar <gregojquiros@gmail.com>
Co-authored-by: Álvaro Rodríguez <sirasistant@gmail.com>
